### PR TITLE
[core] Cap total memory used by executing tasks' arguments

### LINKED ISSUE #13994: [Memory admission control] Implement pinned tasks args threshold.
### Describe your feature request

It seems like when there are lots of pinned_in_memory, the object creation sometimes failed (even with object spilling) because the return objects couldn't be created.

One of the solutions is to introduce the pinned_task args threshold. For example, we can stop scheduling tasks when the pinned task args are occupying more than 70% of object store memory.

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Task arguments must be pinned in memory while the task is running. If too many tasks with arguments run at the same time, there may not be enough memory to store a single task output. Object spilling does not help because the task arguments cannot be spilled while the task is still running.

This adds a cap to the total memory taken by pinned task arguments (default 70% of object store memory). A single task whose arguments are greater than this size will still be scheduled immediately, since otherwise the task would hang or starve.

## Related issue number

Closes #13994 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/scheduling/cluster_task_manager_test.cc [hunk 1]
```diff
@@ -190,18 +192,18 @@ class ClusterTaskManagerTest : public ::testing::Test {
     ASSERT_TRUE(task_manager_.waiting_tasks_index_.empty());
     ASSERT_TRUE(task_manager_.waiting_task_queue_.empty());
     ASSERT_TRUE(task_manager_.infeasible_tasks_.empty());
+    ASSERT_TRUE(task_manager_.executing_task_args_.empty());
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager_test.cc:195]: Should we also check pinned_task_args_bytes is zero?

Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_task_manager_test.cc:195]: Ah thanks, I missed that one.

File: python/ray/tests/test_object_manager.py [hunk 2]
```diff
@@ -375,6 +375,39 @@ def allocate(i):
...
+    @ray.remote
+    def f(arg):
+        time.sleep(1)
+        return np.zeros(30 * 1024 * 1024, dtype=np.uint8)
+
+    # Each task arg takes about 30% of the remote node's memory. We should
+    # execute at most 2 at a time to make sure we have room for at least 1 task
+    # output.
+    x = np.zeros(30 * 1024 * 1024, dtype=np.uint8)
+    ray.get([f.remote(ray.put(x)) for _ in range(3)])
```
Comment (rkooo567) [inline python/ray/tests/test_object_manager.py:399]: This test doesn't really verify that 3 tasks are not running at the same time right?

Why don't we add `Wait for condition used_cpus == 2` before we call ray.get here?

File: src/ray/raylet/scheduling/cluster_task_manager_test.cc [hunk 3]
```diff
@@ -1102,6 +1105,71 @@ TEST_F(ClusterTaskManagerTest, TestSpillWaitingTasks) {
...
+  /* This task can run. */
+  default_arg_size_ = 600;
+  auto task = CreateTask({{ray::kCPU_ResourceLabel, 1}}, 1);
+  task_manager_.QueueAndScheduleTask(task, &reply, callback);
+  ASSERT_EQ(num_callbacks, 1);
+  ASSERT_EQ(leased_workers_.size(), 1);
+  ASSERT_EQ(pool_.workers.size(), 1);
+  AssertPinnedTaskArgumentsPresent(task);
+
+  /* This task cannot run because it would put us over the memory threshold. */
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager_test.cc:1139]: Can you also add a test that another task can run if it shares the same input arg with the first task?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 4]
```diff
@@ -332,15 +323,91 @@ void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
...
+
+void ClusterTaskManager::ReleaseTaskArgs(const TaskID &task_id) {
+  auto it = executing_task_args_.find(task_id);
+  RAY_CHECK(it != executing_task_args_.end());
+  for (auto &arg : it->second) {
+    auto arg_it = pinned_task_arguments_.find(arg);
+    RAY_CHECK(arg_it != pinned_task_arguments_.end());
+    RAY_CHECK(arg_it->second.second > 0);
+    arg_it->second.second--;
+    if (arg_it->second.second == 0) {
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:403]: // We're the last holder of this arg

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 5]
```diff
@@ -332,15 +323,91 @@ void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
...
+}
+
+void ClusterTaskManager::PinTaskArgs(const TaskSpecification &spec,
+                                     std::vector<std::unique_ptr<RayObject>> args) {
+  const auto &deps = spec.GetDependencyIds();
+  for (size_t i = 0; i < deps.size(); i++) {
+    auto inserted =
+        pinned_task_arguments_.emplace(deps[i], std::make_pair(std::move(args[i]), 0));
+    auto it = inserted.first;
+    if (inserted.second) {
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:387]: // We're the first task to pin this arg.

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 6]
```diff
@@ -332,15 +323,91 @@ void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
...
+      }
+    }
+  }
+
+  *args_missing = false;
+  size_t task_arg_bytes = 0;
+  for (auto &arg : args) {
+    task_arg_bytes += arg->GetSize();
+  }
+  PinTaskArgs(spec, std::move(args));
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:364]: It's a bit confusing we have pin_task_args_ and PinTaskArgs, consider renaming one or the other.

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 7]
```diff
@@ -148,44 +150,34 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
...
-        const auto task_id = spec.TaskId();
-        // Insert the task at the head of the waiting queue because we
-        // prioritize spilling from the end of the queue.
-        auto it =
-            waiting_task_queue_.insert(waiting_task_queue_.begin(), std::move(*work_it));
-        RAY_CHECK(waiting_tasks_index_.emplace(task_id, it).second);
-        work_it = dispatch_queue.erase(work_it);
+        if (args_missing) {
+          // Insert the task at the head of the waiting queue because we
+          // prioritize spilling from the end of the queue.
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:163]: Can you explain a bit more about the correlation here (in the comment)? Like why we don't want to spill tasks that have missing args.

**Review by ericl — APPROVED** [review]
LGTM; some clarity and test suggestions

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 7 above)
Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:163]: Also, is inserting to the head of the queue O(N) operation here? Or O(1)?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 8]
```diff
@@ -148,44 +150,34 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
...
+          auto it = waiting_task_queue_.insert(waiting_task_queue_.begin(),
+                                               std::move(*work_it));
+          RAY_CHECK(waiting_tasks_index_.emplace(task_id, it).second);
+          work_it = dispatch_queue.erase(work_it);
+        } else {
+          // The task's args cannot be pinned due to lack of memory.
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:169]: Is it retried right away in this case?

File: src/ray/raylet/scheduling/cluster_task_manager.h [hunk 9]
```diff
@@ -273,12 +274,17 @@ class ClusterTaskManager : public ClusterTaskManagerInterface {
   /// Arguments needed by currently granted lease requests. These should be
   /// pinned before the lease is granted to ensure that the arguments are not
   /// evicted before the task(s) start running.
-  std::unordered_map<TaskID, std::vector<std::unique_ptr<RayObject>>>
+  std::unordered_map<TaskID, std::vector<ObjectID>> executing_task_args_;
+
+  std::unordered_map<ObjectID, std::pair<std::unique_ptr<RayObject>, size_t>>
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.h:279]: Can you comment what this map is for?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 10]
```diff
@@ -332,15 +323,91 @@ void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
...
+}
+
+void ClusterTaskManager::ReleaseTaskArgs(const TaskID &task_id) {
+  auto it = executing_task_args_.find(task_id);
+  RAY_CHECK(it != executing_task_args_.end());
+  for (auto &arg : it->second) {
+    auto arg_it = pinned_task_arguments_.find(arg);
+    RAY_CHECK(arg_it != pinned_task_arguments_.end());
+    RAY_CHECK(arg_it->second.second > 0);
+    arg_it->second.second--;
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:402]: Can you use something like 

```
auto &obj_ref_pair = arg_it->second; 
```

I think `arg_it->second.second` is a bit hard to read.

File: python/ray/tests/test_object_manager.py (hunk 2 above)
Comment (stephanie-wang) [reply python/ray/tests/test_object_manager.py:399]: It does; before this test would error with ObjectStoreFullError.

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 7 above)
Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:163]: This was copy-pasted from the original code. It's an O(1) operation.

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 8 above)
Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:169]: It's retried once another task completes (see the log message) under this.

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**